### ISSUE TITLE
Overlapping blocks

### ISSUE THREAD

**Issue by moliata** [mit-cml/appinventor-sources#1071]
A few months ago, Ewan made a commit 853908e07d2311510b67c9f465cf1f03ac3ce6de which had to fixed #920, but after that commit I didn't see an issue being fix... I have already given a notice about that to Hossein, who made a post on Open Source Development forum.
Indeed, blocks states are not stored during upgrades. This is probably happening because either block positions are stored incorrectly or we do not retain them correctly.

**Comment by ewpatton** [comment]
It is possible there are multiple paths through the code that cause this bug. I fixed it using Mole Mash as an example. Are you seeing the issue if you import the Mole Mash project? Can you provide a list of components in the project that was affected? This might point us in the right direction.

**Comment by kkashi01** [comment]
@ewpatton Haven't really found the cause and the pattern. Its been reported by couple of AppyBuilder members (see below). Its very unpredictable and haven't been able to reproduce

http://community.appybuilder.com/t/something-weird-on-the-blocks/2916?source_topic_id=3387
http://community.appybuilder.com/t/all-blocks-on-same-place/3387

**Comment by ewpatton** [comment]
@kkashi01 Do you have backups turned on? Could you see if you could grab an older version of an effected app from GCS?

**Comment by kkashi01** [comment]
@ewpatton  Could you please let me know how the backup had to be turned on and how to retrieve older version? Not sure if I have seen a doc on this. THANKS

**Comment by ewpatton** [comment]
As far as I know, the backup option is always on--there isn't a configurable way to turn it off short of commenting out the code in ObjectifyStorageIo. The way I typically access the backup content is through the Storage section of the GCP console. Here's an example for one of my test servers:

![image](https://user-images.githubusercontent.com/1331253/34396600-1e98f606-eb3b-11e7-9000-f52a54e70930.png)

Backups happen if the file being modified hasn't been touched for 24 hours, by default.

**Comment by ewpatton** [comment]
See [ObjectifyStorageIo.java:1647](https://github.com/mit-cml/appinventor-sources/blob/77b6a6165b996bc55dde66eb7e0c641a1d492c24/appinventor/appengine/src/com/google/appinventor/server/storage/ObjectifyStorageIo.java#L1647)

**Comment by ewpatton** [comment]
Crossreferencing related issue #961

**Comment by kkashi01** [comment]
@ewpatton Thanks. Let me get permission of user to find out if okay to access send a version of his app

**Comment by ewpatton** [comment]
If you could download the whole src folder that would be helpful. Then we could "replay" and see at what point things went south.

**Comment by ColinTree** [comment]
![image](https://user-images.githubusercontent.com/22613139/34404519-01ac0174-ebe9-11e7-9893-983d12b8ae79.png)
Same to me

**Comment by ewpatton** [comment]
@kkashi01 There isn't a way to search. Assuming you have access to the account, you can get the project ID, which is used for binning the backups as you can observe in the image I attached previously. 

Using the `gsutil` command (part of Google Cloud SDK), you can download the files. From the above example:
```shell
gsutil cp -r gs://ai2-ewpatton-temp.appspot.com/5629499534213120/ .
```

**Comment by ewpatton** [comment]
@ColinTree Was this on ai2? If so, let me know the project ID and we can try and pull it to have a look.

**Comment by ColinTree** [comment]
@ewpatton 
colinycl123@gmail.com
id 4840419484237824

**Comment by moliata** [comment]
@ewpatton after further investigation, I see that blocks do not have specific event that causes this issue, but I can say that when I was testing a local instance of app inventor, I accidently reloaded a webpage and after opening my blocks, this issue occurred. I wasn't been able to reproduce this again.

**Comment by ewpatton** [comment]
I just managed to trigger this as well. At least in my scenario, when I created a new project (so not an upgrade), the blocks XML contains X and Y as undefined, which is clearly wrong. Reading it back into App Inventor converts `undefined` to `0`. Unfortunately, when I tried it to replicate it I was getting correct X and Y values. I think I might be able to track down the issue now knowing that the problem seems to occur in the serialization of the blocks.
